lib/common/schemas.c

```diff
@@ -85,27 +76,54 @@ static int
 xml_latest_schema_index(void)
 {
     // @COMPAT: pacemaker-next is deprecated since 2.1.5
-    return xml_schema_max - 3; // index from 0, ignore "pacemaker-next"/"none"
+    // FIXME: This function assumes at least three schemas have been added
+    // before it has been called for the first time.
+    return g_list_length(known_schemas) - 3; // index from 0, ignore "pacemaker-next"/"none"
 }
 
+/* Return the index of the most recent X.0 schema. */
 static int
-xml_minimum_schema_index(void)
+xml_find_x_0_schema_index(void)
 {
     static int best = 0;
-    if (best == 0) {
-        int lpc = 0;
-
-        best = xml_latest_schema_index();
-        for (lpc = best; lpc > 0; lpc--) {
-            if (known_schemas[lpc].version.v[0]
-                < known_schemas[best].version.v[0]) {
-                return best;
-            } else {
-                best = lpc;
-            }
+    int i = 0;
+    struct schema_s *best_schema = NULL;
+
+    if (best != 0) {
+        return best;
+    }
+
+    /* Get the most recent schema so we can look at its version number. */
+    best = xml_latest_schema_index();
+    best_schema = g_list_nth(known_schemas, best)->data;
+
+    /* Iterate over the schema list until we find a schema with the same major
+     * version as best, and with a minor version number of 0.
+     *
+     * This assumes that the first schema in a major series is always X.0,
+     * which seems like a safe assumption.
```

Comment:
It should be, but we could avoid the assumption by stepping backward with prev() and returning the earliest schema with the same major version -- should be quicker in most cases too. You can do that in a separate PR if desired.

Reply:
See #3262.  This would be great to have a unit test for, but I don't want to make the function and the struct public.